File: app/src/main/java/com/example/crudifyapplication/ProductListAdapter.java
package com.example.crudifyapplication;

import android.content.Context;
import android.database.Cursor;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.crudifyapplication.R;

public class ProductListAdapter extends RecyclerView.Adapter<ProductListAdapter.ViewHolder> {

    private Context context;
    private Cursor cursor;
    private OnProductActionListener actionListener;

    // Constructor with context, cursor, and action listener
    public ProductListAdapter(Context context, Cursor cursor, OnProductActionListener actionListener) {
        this.context = context;
        this.cursor = cursor;
        this.actionListener = actionListener;
    }

    // Define the interface for edit and delete actions
    public interface OnProductActionListener {
        void onEditClick(int position, int productId);
        void onDeleteClick(int position, int productId);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.product_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        if (!cursor.moveToPosition(position)) {
            return;
        }

        // Retrieve data from cursor
        int id = cursor.getInt(cursor.getColumnIndexOrThrow("id"));
        String name = cursor.getString(cursor.getColumnIndexOrThrow("name"));
        int quantity = cursor.getInt(cursor.getColumnIndexOrThrow("quantity"));

        // Bind data to the views in the ViewHolder
        holder.productIdTextView.setText(String.valueOf(id));
        holder.productNameTextView.setText(name);
        holder.productQuantityTextView.setText(String.valueOf(quantity));

        // Set listeners for edit and delete icons
        holder.editIcon.setOnClickListener(v -> actionListener.onEditClick(position, id));
        holder.deleteIcon.setOnClickListener(v -> actionListener.onDeleteClick(position, id));
    }

    @Override
    public int getItemCount() {
        return cursor != null ? cursor.getCount() : 0;
    }

    // ViewHolder class to hold references to each item’s views
    public static class ViewHolder extends RecyclerView.ViewHolder {
        TextView productIdTextView, productNameTextView, productQuantityTextView;
        ImageView editIcon, deleteIcon;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            productIdTextView = itemView.findViewById(R.id.productIdTextView);
            productNameTextView = itemView.findViewById(R.id.productNameTextView);
            productQuantityTextView = itemView.findViewById(R.id.productQuantityTextView);
            editIcon = itemView.findViewById(R.id.editIcon);
            deleteIcon = itemView.findViewById(R.id.deleteIcon);
        }
    }

    // Method to update the cursor when data changes
    public void swapCursor(Cursor newCursor) {
        if (cursor != null) {
            cursor.close();
        }
        cursor = newCursor;
        if (newCursor != null) {
            notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/com/example/crudifyapplication/adapter/ProductListAdapter.kt
package com.example.crudifyapplication.adapter

import android.content.Context
import android.database.Cursor
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.crudifyapplication.R

class ProductListAdapter(
    private val context: Context,
    private var cursor: Cursor,
    private val listener: OnProductActionListener
) : RecyclerView.Adapter<ProductListAdapter.ProductViewHolder>() {

    interface OnProductActionListener {
        fun onEditClick(position: Int, productId: Int)
        fun onDeleteClick(position: Int, productId: Int)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.product_item, parent, false)
        return ProductViewHolder(view)
    }

    override fun onBindViewHolder(holder: ProductViewHolder, position: Int) {
        if (cursor.moveToPosition(position)) {
            val productId = cursor.getInt(cursor.getColumnIndexOrThrow("product_id"))
            val productName = cursor.getString(cursor.getColumnIndexOrThrow("product_name"))
            val productQuantity = cursor.getInt(cursor.getColumnIndexOrThrow("product_quantity"))

            holder.productIdTextView.text = productId.toString()
            holder.productNameTextView.text = productName
            holder.productQuantityTextView.text = productQuantity.toString()

            holder.editIcon.setOnClickListener {
                listener.onEditClick(position, productId)
            }

            holder.deleteIcon.setOnClickListener {
                listener.onDeleteClick(position, productId)
            }
        }
    }

    override fun getItemCount(): Int = cursor.count

    fun swapCursor(newCursor: Cursor?) {
        if (cursor != newCursor) {
            val oldCursor = cursor
            cursor = newCursor ?: cursor
            notifyDataSetChanged()
            oldCursor?.close()
        }
    }

    class ProductViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val productIdTextView: TextView = itemView.findViewById(R.id.productIdTextView)
        val productNameTextView: TextView = itemView.findViewById(R.id.productNameTextView)
        val productQuantityTextView: TextView = itemView.findViewById(R.id.productQuantityTextView)
        val editIcon: ImageView = itemView.findViewById(R.id.editIcon)
        val deleteIcon: ImageView = itemView.findViewById(R.id.deleteIcon)
    }
}
